src/hitrecord.nim

```diff
@@ -1,201 +1,161 @@
 
-        if tspan.min > worldInvRay.tspan.max or tspan.max < worldInvRay.tspan.min: return none HitPayload
+        if tspan.min > worldInvRay.tspan.max or tspan.max < worldInvRay.tspan.min: return Inf
 
-        var tHit = tspan.min
-        if tHit < worldInvRay.tspan.min:
-            if tspan.max > worldInvRay.tspan.max: return none HitPayload
-            tHit = tspan.max
+        result = tspan.min
+        if result < worldInvRay.tspan.min:
+            if tspan.max > worldInvRay.tspan.max: return Inf
+            result = tspan.max
 
-        var hitPt = worldInvRay.at(tHit)
+        var hitPt = worldInvRay.at(result)
         var phi = arctan2(hitPt.y, hitPt.x)
         if phi < 0.0: phi += 2.0 * PI
 
-        if hitPt.z < handler.shape.zSpan.min or hitPt.z > handler.shape.zSpan.max or phi > handler.shape.phiMax:
-            if tHit == tspan.max: return none HitPayload
-            tHit = tspan.max
-            if tHit > worldInvRay.tspan.max: return none HitPayload
+        if hitPt.z < shape.zSpan.min or hitPt.z > shape.zSpan.max or phi > shape.phiMax:
+            if result == tspan.max: return Inf
+            result = tspan.max
+            if result > worldInvRay.tspan.max: return Inf
             
-            hitPt = worldInvRay.at(tHit)
+            hitPt = worldInvRay.at(result)
             phi = arctan2(hitPt.y, hitPt.x)
             if phi < 0.0: phi += 2.0 * PI
-            if hitPt.z < handler.shape.zSpan.min or hitPt.z > handler.shape.zSpan.max or phi > handler.shape.phiMax: return none HitPayload
-
-        return some HitPayload(handler: handler, ray: worldInvRay, t: tHit)
+            if hitPt.z < shape.zSpan.min or hitPt.z > shape.zSpan.max or phi > shape.phiMax: return Inf
 
-    of skTriangularMesh: 
-        let localHitLeafNodes = handler.shape.tree.getHitLeafs(worldInvRay)
-        if localHitLeafNodes.isNone: return none HitPayload
-        
-        proc localHitPayloads(sceneTree: SceneNode; worldInvRay: Ray): seq[HitPayload] =
-            var hittedHandlers: seq[ShapeHandler]
-            for handler in sceneTree.handlers:
-                if checkIntersection(handler.getAABB, worldInvRay): hittedHandlers.add handler
 
-            if hittedHandlers.len == 0: return @[]
+type HitInfo[T] = tuple[hit: T, t: float32]
 
-            hittedHandlers
-                .mapIt(it.getHitPayload(worldInvRay))
-                .filterIt(it.isSome)
-                .mapIt(it.get)
+proc getClosestHit*(tree: BVHTree, handlers: seq[ObjectHandler], worldRay: Ray): HitInfo[ObjectHandler] =
+    result = (nil, Inf.float32)
+    if tree.root.isNil: return result
 
-        proc localHitRecord(hitLeafs: seq[SceneNode]; worldInvRay: Ray): Option[seq[HitPayload]] =
-            let hitPayloads = hitLeafs
-                .mapIt(it.localHitPayloads(worldInvRay))
-                .filterIt(it.len > 0)
+    var nodeStack: seq[HitInfo[BVHNode]] = @[(tree.root, tree.root.aabb.getIntersection(worldRay))]
```

Comment:
I feel like it could be a good idea to check if you have intersection with tree root aabb, because then it's no sense to check children nodes